### PR TITLE
Add previous implementation

### PR DESCRIPTION
**Pull request by ulyfm**

### CODE REVIEW

File: src/main/java/StringManipulation.java [hunk 1]
```diff
@@ -1,32 +1,79 @@
 public class StringManipulation implements StringManipulationInterface {
+    private String myString = "";
```
Comment (BladeTheHunter005) [inline src/main/java/StringManipulation.java:3]: Great job creating the StringManipulation class and implementing the StringManipulation Interface.

File: src/main/java/StringManipulation.java [hunk 2]
```diff
@@ -1,32 +1,79 @@
...
 
     @Override
     public String getString() {
-        return null;
+        return myString;
     }
```
Comment (BladeTheHunter005) [inline src/main/java/StringManipulation.java:7]: It's good to see the getString() method implemented, which returns the value of myString. It allows users to retrieve the current string.

File: src/main/java/StringManipulation.java [hunk 3]
```diff
@@ -1,32 +1,79 @@
...
 
     @Override
     public void setString(String string) {
+        myString = string;
     }
```
Comment (BladeTheHunter005) [inline src/main/java/StringManipulation.java:12]: Method effectively sets the value of myString to the provided string.

File: src/main/java/StringManipulation.java [hunk 4]
```diff
@@ -1,32 +1,79 @@
...
-        return 0;
+        String[] spl = myString.split("\\s+");
+        int i = 0;
+        for (String str : spl) {
+            if (!str.equals("")) {
+                i++;
+            }
+        }
+        return i;
     }
```
Comment (BladeTheHunter005) [inline src/main/java/StringManipulation.java:24]: Have you considered using the enhanced for-loop syntax for the count() method to improve readability? For example, 
for(String str: spl) instead of using traditional for-loop.

To make the code even more readable, could you add a comment explaining what the purpose of the count() method is?
I know we all know what it is and what it does, but good practice is documenting everything.

File: src/main/java/StringManipulation.java [hunk 5]
```diff
@@ -1,32 +1,79 @@
...
+            if (i != 0 && i % n == 0) {
+                if (maintainSpacing) {
+                    ret.append(" ");
+                }
+            } else {
+                ret.append(ar[i - 1]);
+            }
+        }
+        return ret.toString();
     }
```
Comment (BladeTheHunter005) [inline src/main/java/StringManipulation.java:45]: This removeNthCharacter(int n, boolean maintainSpacing) method looks good.
It is good to see you handling cases where n is greater than myString.length() and when n is less than or equal to 0.

Once again have you consider commenting detailed documentation so that your code is easy to modify. Remember in the verification and validation phase of testing that we have to find bugs, ensure readability. Consider documenting everything.

File: src/main/java/StringManipulation.java [hunk 6]
```diff
@@ -1,32 +1,79 @@
...
+        String[] ret = new String[endWord - startWord + 1];
+        String[] spl = myString.split("\\s+");
+        if (spl.length < endWord) {
+            throw new IndexOutOfBoundsException();
+        }
+        for (int i = 0; i < ret.length; i++) {
+            ret[i] = spl[startWord - 1 + i];
+        }
+        return ret;
     }
```
Comment (BladeTheHunter005) [inline src/main/java/StringManipulation.java:61]: Great job validating input parameters in the getSubStrings(int startWord, int endWord) method. to ensure they met necessary conditions.

File: src/main/java/StringManipulation.java [hunk 7]
```diff
@@ -1,32 +1,79 @@
...
+        StringBuilder ret = new StringBuilder();
+        char[] ar = myString.toCharArray();
+        for (int i = 0; i < indices.length; i++) {
+            if (indices[i] < 0 || indices[i] >= ar.length) {
+                throw new IndexOutOfBoundsException();
+            }
+            ret.append(ar[indices[i]]);
+        }
+        return ret.toString();
     }
```
Comment (BladeTheHunter005) [inline src/main/java/StringManipulation.java:77]: Excellent job validating the indices array in the restoreString(int[] indices) method to ensure it has the same length as myString.

File: src/test/java/StringManipulationTest.java [hunk 8]
```diff
@@ -2,7 +2,6 @@
 import org.junit.jupiter.api.BeforeEach;
 import org.junit.jupiter.api.Test;
```
Comment (BladeTheHunter005) [inline src/test/java/StringManipulationTest.java:4]: Great job on importing necessary libraries for your tests. It's clear that you've thought about what you need.

File: src/test/java/StringManipulationTest.java [hunk 9]
```diff
@@ -19,65 +18,109 @@ public void tearDown() {
...
 
+    /**
+     * Check whether a simple string of space-separated words is counted correctly.
+     */
     @Test
     public void testCount1() {
         manipulatedstring.setString("This is my string");
         int length = manipulatedstring.count();
         assertEquals(4, length);
     }
```
Comment (BladeTheHunter005) [inline src/test/java/StringManipulationTest.java:29]: Good job on testing for an empty string in testCount2. This is a good edge case to consider.

File: src/test/java/StringManipulationTest.java [hunk 10]
```diff
@@ -19,65 +18,109 @@ public void tearDown() {
...
 
+    /**
+     * Check whether a single letter string is correctly counted as one.
+     */
     @Test
     public void testCount3() {
-        fail("Not yet implemented");
+        manipulatedstring.setString("A");
+        assertEquals(1, manipulatedstring.count());
     }
```
Comment (BladeTheHunter005) [inline src/test/java/StringManipulationTest.java:47]: You are testing a single character string. This is another good edge case. Have you considered testing a string with only one space.

File: src/test/java/StringManipulationTest.java [hunk 11]
```diff
@@ -19,65 +18,109 @@ public void tearDown() {
...
 
+    /**
+     * Check whether a string of multiple spaces is correctly counted as zero.
+     */
     @Test
     public void testCount4() {
-        fail("Not yet implemented");
+        manipulatedstring.setString("   ");
+        assertEquals(0, manipulatedstring.count());
     }
```
Comment (BladeTheHunter005) [inline src/test/java/StringManipulationTest.java:56]: what do you think would happen if the string contained tabs or newline characters?

File: src/test/java/StringManipulationTest.java [hunk 12]
```diff
@@ -19,65 +18,109 @@ public void tearDown() {
...
 
+    /**
+     * Check whether every third character is correctly removed (not maintaining spacing) in a simple scenario.
+     */
     @Test
     public void testRemoveNthCharacter1() {
         manipulatedstring.setString("I'd b3tt3r put s0me d161ts in this 5tr1n6, right?");
         assertEquals("I' bttr uts0e 16tsinths trn6 rgh?", manipulatedstring.removeNthCharacter(3, false));
     }
```
Comment (BladeTheHunter005) [inline src/test/java/StringManipulationTest.java:65]: testing the removal of the third character is a good test case.

File: src/test/java/StringManipulationTest.java [hunk 13]
```diff
@@ -87,27 +130,66 @@ public void testGeSubStrings1() {
         assertEquals(sStings[1], "string");
     }
```
Comment (BladeTheHunter005) [inline src/test/java/StringManipulationTest.java:131]: you're testing the splitting of a string into substrings. This is a good test case. Have you considered what might happen if the string contains multiple spaces between words?

File: src/test/java/StringManipulationTest.java [hunk 14]
```diff
@@ -87,27 +130,66 @@ public void testGeSubStrings1() {
...
+     * Check whether a simple string with words separated by multiple spaces is correctly converted to substrings.
+     */
+    @Test
+    public void testGeSubStrings7() {
+        manipulatedstring.setString("This    is    my   string");
+        String [] sStings = manipulatedstring.getSubStrings(3, 4);
+        assertEquals(sStings[0], "my");
+        assertEquals(sStings[1], "string");
+    }
+
```
Comment (BladeTheHunter005) [inline src/test/java/StringManipulationTest.java:189]: This is a good test case. Have you considered what might happen if the string contains other types of whitespace characters, like tabs or newlines?

File: src/test/java/StringManipulationTest.java [hunk 15]
```diff
@@ -118,32 +200,51 @@ public void testRestoreString1()
...
+        int [] array = new int[]{1,2,3};
+        assertThrows(IndexOutOfBoundsException.class, () -> manipulatedstring.restoreString(array));
     }
 
+    /**
+     * Check whether a string with non-alphanumeric characters can be successfully rearranged. (Also not part
+     * of the spec, but my implementation does it anyway and I wanted to test it.)
+     */
     @Test
     public void testRestoreString5()
```
Comment (BladeTheHunter005) [inline src/test/java/StringManipulationTest.java:243]: And this one, you're throwing in non-alphanumeric characters into the mix. Real-world strings can be messy and full of all sorts of characters, so it's awesome that you're making sure your method can handle that

File: src/test/java/StringManipulationTest.java [hunk 16]
```diff
@@ -118,32 +200,51 @@ public void testRestoreString1()
...
+        int [] array = new int[]{2,2,2};
+        assertEquals(manipulatedstring.restoreString(array), "ccc");
     }
 
+    /**
+     * Check whether an illegal argument exception is correctly thrown when the number of indices doesn't match
+     * the number of letters in the string.
+     */
     @Test
     public void testRestoreString3()
```
Comment (BladeTheHunter005) [inline src/test/java/StringManipulationTest.java:220]: Here, you're testing if your method can deal with a mismatch in the number of indices and the number of characters in the string. It's like trying to fit a square peg in a round hole - it's not gonna work, and your method should recognize that. Good on you for checking this!

**Review by BladeTheHunter005 — APPROVED** [review]
Project accepted. Code review successfully completed
